File: project/drivers/driver/go/readiness_test.go
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

package scratchbird

import (
	"database/sql"
	"testing"
)

func TestBetaDriverReadinessStatusMatchesManifestLane(t *testing.T) {
	status := BetaDriverReadinessStatus()
	if status.ComponentID != "driver:go" {
		t.Fatalf("unexpected component id: %s", status.ComponentID)
	}
	if status.DriverPackageUUID != "019e12a0-0005-7000-8000-000000000005" {
		t.Fatalf("unexpected package uuid: %s", status.DriverPackageUUID)
	}
	if status.DriverStatus != "beta_2" || status.ReleaseBucket != "release_candidate" {
		t.Fatalf("unexpected manifest status: %#v", status)
	}
	if !status.AuthorityBoundary.ServerRevalidationRequired {
		t.Fatalf("server revalidation must remain required")
	}
	if !status.AuthorityBoundary.LocalSBLRIsAdvisory || !status.AuthorityBoundary.LocalUUIDCacheIsAdvisory {
		t.Fatalf("local SBLR/UUID cache must be advisory")
	}
	if status.AuthorityBoundary.TransactionFinalityOwner != "engine_mga_transaction_inventory" {
		t.Fatalf("unexpected finality owner: %s", status.AuthorityBoundary.TransactionFinalityOwner)
	}
}

func TestAdvisoryCacheContextRefusesStaleEpochsAndTransactionReuse(t *testing.T) {
	current := AdvisoryCacheContext{
		DatabaseUUID:    "db-1",
		SchemaEpoch:     "schema-2",
		PolicyEpoch:     "policy-2",
		LanguageEpoch:   "lang-2",
		CapabilityEpoch: "cap-2",
		PrincipalUUID:   "principal-1",
		RoleSetHash:     "role-a",
		GroupSetHash:    "group-a",
		TransactionUUID: "txn-2",
	}
	cached := current
	cached.PolicyEpoch = "policy-1"
	diag := ValidateAdvisoryCacheContext(cached, current)
	if diag == nil || diag.Code != "SB_DRIVER_CACHE_POLICY_EPOCH_STALE" {
		t.Fatalf("expected stale policy diagnostic, got %#v", diag)
	}

	cached = current
	cached.TransactionUUID = "txn-1"
	diag = ValidateAdvisoryCacheContext(cached, current)
	if diag == nil || diag.Code != "SB_DRIVER_CACHE_TRANSACTION_CONTEXT_MISMATCH" || diag.SQLState != "25001" {
		t.Fatalf("expected transaction boundary diagnostic, got %#v", diag)
	}

	if diag = ValidateAdvisoryCacheContext(current, current); diag != nil {
		t.Fatalf("matching advisory context should pass, got %#v", diag)
	}
}

func TestPreparedBundleReuseRequiresServerAdmissionAndMatchingContext(t *testing.T) {
	current := AdvisoryCacheContext{
		DatabaseUUID:    "db-1",
		SchemaEpoch:     "schema-1",
		PolicyEpoch:     "policy-1",
		PrincipalUUID:   "principal-1",
		TransactionUUID: "txn-1",
	}
	bundle := PreparedBundleContext{
		DatabaseUUID:    "db-1",
		SchemaEpoch:     "schema-1",
		PolicyEpoch:     "policy-1",
		PrincipalUUID:   "principal-1",
		TransactionUUID: "txn-1",
		ServerAdmitted:  false,
	}
	diag := ValidatePreparedBundleReuse(bundle, current)
	if diag == nil || diag.Code != "SB_DRIVER_SBLR_SERVER_ADMISSION_REQUIRED" {
		t.Fatalf("expected server admission diagnostic, got %#v", diag)
	}

	bundle.ServerAdmitted = true
	bundle.DatabaseUUID = "db-2"
	diag = ValidatePreparedBundleReuse(bundle, current)
	if diag == nil || diag.Code != "SB_DRIVER_CACHE_DATABASE_MISMATCH" {
		t.Fatalf("expected cross-database refusal, got %#v", diag)
	}
}

func TestLanguageResourceFallbackAndIntegrityRefusal(t *testing.T) {
	resolved := ResolveLanguageProfile("fr_CA", map[string]bool{"en_US": true})
	if resolved.Selected != "en_US" || !resolved.Fallback {
		t.Fatalf("expected standard English fallback, got %#v", resolved)
	}

	resolved = ResolveLanguageProfile("fr_CA", map[string]bool{"en_US": true, "fr_CA": true})
	if resolved.Selected != "fr_CA" || resolved.Fallback {
		t.Fatalf("expected preferred language selection, got %#v", resolved)
	}

	diag := ValidateLanguageResourceState(LanguageResourceState{
		Locale:        "fr_CA",
		SchemaVersion: "1",
		ContentHash:   "sha256:abc",
		Signature:     "sig",
		Epoch:         "lang-1",
		ExpectedEpoch: "lang-2",
	})
	if diag == nil || diag.Code != "SB_DRIVER_LANGUAGE_RESOURCE_EPOCH_STALE" {
		t.Fatalf("expected stale language resource diagnostic, got %#v", diag)
	}
}

func TestPackageSmokeRegistersDatabaseSQLDriver(t *testing.T) {
	for _, name := range sql.Drivers() {
		if name == "scratchbird" {
			return
		}
	}
	t.Fatalf("scratchbird database/sql driver was not registered")
}

File: project/drivers/driver/rust/tests/readiness_test.rs
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

use scratchbird::{
    beta_driver_readiness_status, resolve_language_profile, validate_advisory_cache_context,
    validate_language_resource_state, validate_prepared_bundle_reuse, AdvisoryCacheContext,
    LanguageResourceState, PreparedBundleContext,
};

fn current_context() -> AdvisoryCacheContext {
    AdvisoryCacheContext {
        database_uuid: "db-1".to_string(),
        schema_epoch: "schema-2".to_string(),
        policy_epoch: "policy-2".to_string(),
        language_epoch: "lang-2".to_string(),
        capability_epoch: "cap-2".to_string(),
        principal_uuid: "principal-1".to_string(),
        role_set_hash: "role-a".to_string(),
        group_set_hash: "group-a".to_string(),
        transaction_uuid: "txn-2".to_string(),
    }
}

#[test]
fn beta_driver_readiness_status_matches_manifest_lane() {
    let status = beta_driver_readiness_status();
    assert_eq!(status.component_id, "driver:rust");
    assert_eq!(
        status.driver_package_uuid,
        "019e12a0-0015-7000-8000-000000000015"
    );
    assert_eq!(status.driver_status, "beta_2");
    assert_eq!(status.release_bucket, "release_candidate");
    assert_eq!(status.conformance_profile_ref, "driver_rust_gate");
    assert!(status.authority_boundary.server_revalidation_required);
    assert!(status.authority_boundary.local_sblr_is_advisory);
    assert!(status.authority_boundary.local_uuid_cache_is_advisory);
    assert_eq!(
        status.authority_boundary.transaction_finality_owner,
        "engine_mga_transaction_inventory"
    );
    assert!(status.runtime_mapping.dsn_keys.contains(&"auth_method"));
}

#[test]
fn advisory_cache_context_refuses_stale_epochs_and_transaction_reuse() {
    let current = current_context();
    let mut cached = current.clone();
    cached.policy_epoch = "policy-1".to_string();
    let diag = validate_advisory_cache_context(&cached, &current).expect("stale policy refusal");
    assert_eq!(diag.code, "SB_DRIVER_CACHE_POLICY_EPOCH_STALE");
    assert_eq!(diag.sqlstate, "42501");

    let mut cached = current.clone();
    cached.transaction_uuid = "txn-1".to_string();
    let diag =
        validate_advisory_cache_context(&cached, &current).expect("transaction boundary refusal");
    assert_eq!(diag.code, "SB_DRIVER_CACHE_TRANSACTION_CONTEXT_MISMATCH");
    assert_eq!(diag.sqlstate, "25001");
    assert!(validate_advisory_cache_context(&current, &current).is_none());
}

#[test]
fn prepared_bundle_reuse_requires_server_admission_and_matching_context() {
    let current = AdvisoryCacheContext {
        database_uuid: "db-1".to_string(),
        schema_epoch: "schema-1".to_string(),
        policy_epoch: "policy-1".to_string(),
        principal_uuid: "principal-1".to_string(),
        transaction_uuid: "txn-1".to_string(),
        ..AdvisoryCacheContext::default()
    };
    let mut bundle = PreparedBundleContext {
        database_uuid: "db-1".to_string(),
        schema_epoch: "schema-1".to_string(),
        policy_epoch: "policy-1".to_string(),
        principal_uuid: "principal-1".to_string(),
        transaction_uuid: "txn-1".to_string(),
        server_admitted: false,
    };
    let diag = validate_prepared_bundle_reuse(&bundle, &current).expect("admission refusal");
    assert_eq!(diag.code, "SB_DRIVER_SBLR_SERVER_ADMISSION_REQUIRED");

    bundle.server_admitted = true;
    bundle.database_uuid = "db-2".to_string();
    let diag = validate_prepared_bundle_reuse(&bundle, &current).expect("database refusal");
    assert_eq!(diag.code, "SB_DRIVER_CACHE_DATABASE_MISMATCH");
}

#[test]
fn language_resource_fallback_and_integrity_refusal() {
    let resolved = resolve_language_profile("fr_CA", &["en_US"]);
    assert_eq!(resolved.selected, "en_US");
    assert!(resolved.fallback);

    let resolved = resolve_language_profile("fr_CA", &["en_US", "fr_CA"]);
    assert_eq!(resolved.selected, "fr_CA");
    assert!(!resolved.fallback);

    let diag = validate_language_resource_state(&LanguageResourceState {
        locale: "fr_CA".to_string(),
        schema_version: "1".to_string(),
        content_hash: "sha256:abc".to_string(),
        signature: "sig".to_string(),
        epoch: "lang-1".to_string(),
        expected_epoch: "lang-2".to_string(),
    })
    .expect("stale language refusal");
    assert_eq!(diag.code, "SB_DRIVER_LANGUAGE_RESOURCE_EPOCH_STALE");
}

#[test]
fn package_smoke_reports_mpl_license() {
    assert_eq!(env!("CARGO_PKG_NAME"), "scratchbird");
    assert_eq!(env!("CARGO_PKG_LICENSE"), "MPL-2.0");
}
